### ISSUE TITLE
Move away from HTML5 hashbang mode

### ISSUE THREAD

**Issue by HarelM** [IsraelHikingMap/Site#653]
## What I expect to happen
URLs should be visible to crawlers/google. in order to do so all URLs should be converted to regular URLs and not use hashbang mode (.../#!/...)

## What really happened
All site URLs use # mode 

## Thing I think the developers should know, images, links etc.
There are many things to consider when solving this:
1. All old URLs should still work but should be redirected to new URLs
2. Server side should know which URLs to handle and which it shouldn't (for example /oruxmaps/ and other file sharing URLs should still work as they work now.
3. The server should be able to see URLs and agent to route crawlers to the right controller to produce the required images.
4. This is not an easy implementation and I couldn't find an article talking about how to do it correctly up until now.

**Comment by SafwatHalaby** [comment]
I think there is one advantage to hasbangs over regular URLs in single page applications. use the  which I've never seen discussed online. Please correct me if I'm wrong. I think Hashbangs are better for local browser caching.

The "app" is made of static content (code, etc) and data is fetched from the server. With hashbangs, the static content is only fetched exactly once (assuming proper not-modified headers).

With changing urls, the initial html page is fetched whenever a new url is landed it.

The only online discussion I've ever found is here:

https://stackoverflow.com/questions/27356774/is-pushstate-inferior-to-hashbangs-when-it-comes-to-caching

**Comment by HarelM** [comment]
I'm not an expert on this subject, but you might have a point as it sounds reasonable.
In any case, I don't see me getting into this in the near future as it requires a lot of work and the benefit from it is low...

**Comment by HarelM** [comment]
#513 will use URL for sharing and we probably need to decide if it will use hashbang or pushstate before publishing it, and so this issue has a higher priority now.
I'll try to summarize pros and cons of the migration:
Pros:
1. Prettier URLs (not really interesting)
2. Allow whatsapp to present the right info when sharing - probable the main reason.
3. Better crawlable support (?)
4. Make navigation an integrated part of the site so that every screen that opens is defined in the address for easier sharing - need to think how to build the navigation addresses right as this is not a straight forward implementation.
Cons:
1. Cache (?)
2. if we don't migrate we don't need to change any thing, if we do migrate we'll need to redirect all old URLs and make sure they work - probably the main reason not to.

As a side note, I don't think the site will be able to be crawlable regardless of the implementation since the site doesn't have links like other sites do, the proper solution will probably be solved here: #699.

Some ideas can be seen here at komoot - here's the address of a route with a waypoint open:
https://www.komoot.com/plan/tour/d01AelM0QISfx8=ExnsBiEdr1gmWqoyAA==/@31.9769017,34.7944736,15z?waypoint=poi:d91hd
Changes to the address bar are visible on every change in the content in the page - the address does not follow any standards I'm familiar with.

**Comment by HarelM** [comment]
After further reading, I think that cache shouldn't be an issue:
https://webmasters.stackexchange.com/questions/65694/is-this-way-of-using-pushstate-seo-friendly
Basically the page that get download is the index page which is very small. js files which are the large files are cached and so I think this is a none issue.

**Comment by zstadler** [comment]
> After further reading, I think that cache shouldn't be an issue:
https://webmasters.stackexchange.com/questions/65694/is-this-way-of-using-pushstate-seo-friendly

The first line of this solution _"tells the web crawlers that this is an ajax page and that they should fetch a prerendered page somewhere else"_, based on the [Google recommendation here](https://developers.google.com/webmasters/ajax-crawling/docs/getting-started). However, this page says that the recommendation is officially deprecated as of October 2015 and will not be supported after the summer of 2018.

However, it seems like there Google will be using a different Ajax crawling scheme, base on Google-side rendering, as described in [this blog](https://webmasters.googleblog.com/2017/12/rendering-ajax-crawling-pages.html). The blog also explains how the site is crawled, and how to verify that the crawling works as expected.

**Comment by HarelM** [comment]
The site was missing `<meta name="fragment" content="!">` in the header, which was suspected to cause WhatsApp not to render the data correctly. Adding it doesn't seem to help. So it seems like WhatsApp crawler do not support this deprecated protocol - which is inline with a feature suggestion I sent them to support this protocol.

**Comment by HarelM** [comment]
Proposed routes mapping:
```
/#!/?q=123 => /search/123
/#!/?s=456 => /share/456
/#!/?url=789=> /file/789 or /externalurl/789
/#!/?p=111 => /poi/111
/#!/?downalod => /download/
/#!/7/32/35 => /?location=7,32,35 or /?location=32,35,7 or something else...
```
`baselayer=<layer name or address>` parameter should be available for url, search and maybe POI since they can be embedded.

Static files should still work after this, i.e. `Tiles/*`, `OruxMaps/*` etc.
Let me know if you think it's too verbose or if you have other suggestions.

**Comment by HarelM** [comment]
Can't argue about the search, but the `url` is not really an action it is a content, external one, but still content.
The logic behind it is that currently, and I think I would like to keep it, share, url and location are mutually exclusive, meaning the address bar has either a location, a share, or an external url, which indicate they are different states.
In theory we can also have
`/#!/7/32/35 => /map/7,32,35`.
As a side note, using "/" inside the address bar is just asking for bugs :-) we've used it up until now but I also wrote a proprietary url router instead of using Angular's which had some nasty bugs in it :-/

To summarize my suggestions:
 ```
/#!/?q=123 => /?q=123
/#!/?s=456 => /share/456
/#!/?url=http%3A%2F%2Fabc.com%2F789.gpx => /file/http%3A%2F%2Fabc.com%2F789.gpx or /externalurl/http%3A%2F%2Fabc.com%2F789.gpx - we still need to decide which, I don't want to support both
/#!/?p=111 => /poi/111
/#!/?downalod => /download/
/#!/7/32/35 => /map/7,32,35 or we can do /map/?zoom=7&lat=32&lon=35 which is more in the HTML spirit.
```
We can also keep the `map` after the # to indicate that it's not interesting to the server, but we'll probably need it in the future so I prefer to remove the # completely.

**Comment by zstadler** [comment]
I agree that using path suffixes for mutually-exclusive approaches to the site is very valuable. More valuable than the distinction between "contents" and "action" I proposed. 

Isn't "search" also mutually exclusive with "share" and "external url"? I'm not sure there is value in their combinations.
BTW, they would look like `/share/456?q=123` or `/url/http%3A%2F%2Fabc.com%2F789.gpx?q=123`

I definitely prefer `/url/...` over both `/file/...` and `/externalurl/...`

`/map/7,32,35` seems better to me than `/map/?zoom=7&lat=32&lon=35`.

**Comment by HarelM** [comment]
>Isn't "search" also mutually exclusive with "share" and "external url"? I'm not sure there is value in their combinations.

Sure - no reason to allow search with anything else.
>`/map/7,32,35`

URLs protocol doesn't support `,` so setting `,` will require manual parsing which is redundant. I'll see if I can define a state for the map like `/map/7/32/35` since it's mutually exclusive it might work.
Summarize:
```
/#!/?q=123 => /search/123 or /q/123 or /query/123 - I prefer search, don't see a reason to make it short.
/#!/?s=456 => /share/456
/#!/?url=....gpx?baselayer=... => /url/....gpx?baselayer=...
/#!/?p=111?baselayer=... => /poi/111?baselayer=...
/#!/?downalod => /download/
/#!/7/32/35 => /map/7/32/35 hopefully this is possible.
```

**Comment by HarelM** [comment]
Updated my comment above `map/...` is possible.
Search is the only open issue now.

**Comment by HarelM** [comment]
Final decision:
```
/#!/?q=123 => /search/123
/#!/?s=456 => /share/456
/#!/?url=....gpx?baselayer=... => /url/....gpx?baselayer=...
/#!/?p=OSM__111?baselayer=... => /poi/OSM/111?baselayer=...
/#!/?downalod => /download/
/#!/7/32/35 => /map/7/32/35
```

**Comment by HarelM** [comment]
There might be some small improvements to navigation that can be made, but the feature is mostly implemented. If there will be a request to improve navigation in the future a relevant issue should be opened.
Closing...
